Add sample code showing how to create a new wxEvent class

PBfordev opened the pull request:
Based on [this comment](https://github.com/wxWidgets/wxWidgets/pull/1586#discussion_r331744773) in #1586, I have tried to extend the event sample by showing how to create and use a new `wxEvent`-derived class.

My objective was to create a meaningful event while keeping the code needed for it as short and simple as possible. I am not sure I succeeded at either, so if anyone reviews this, please do not hesitate to shoot this down if you think it is not actually useful. There is a TODO comment in the header file where I did not know what should be there. I did not even attempt to rebake the files at this stage.

When working with the event sample I noticed some issues such as a comment above  `wxDECLARE_EVENT_TABLE();` reading "any class wishing to process wxWidgets events must use this macro" or the log in the gestures frame consuming all the messages. So perhaps the event sample should be cleaned up instead...

PBfordev committed:
Add sample code showing how to create a new wxEvent class

In the event sample, show how create and use a new
wxEvent-derived class.

PBfordev committed:
Clean up the code

PBfordev committed:
Reorganize the code

Put the new event class ChessBoardEvent
in the focus of the example code
while making the helper class ChessBoard
much less visible.

Move wxDEFINE_EVENT() macros from
the header to the source file.

vadz commented:
I like the new example, thanks!

The only thing I might want to change would be to use `wxLog` and log window instead of message boxes from the handlers to make it less annoying to dismiss all those message dialogs.

I guess the link errors in the CI builds are due to bakefile not having been rerun after the changes in the sample bkl file?

PBfordev commented:
> The only thing I might want to change would be to use `wxLog` and log window instead of message boxes from the handlers to make it less annoying to dismiss all those message dialogs.

I do not like how it is handled in the sample. As I wrote above, the gestures frame (understandably) creates an `wxTextCtrl`-based log target which then consumes all the messages in the sample; those may be quite difficult to notice when this frame is floating somewhere. That is why I used (1) the modal dialog and (2) `wxMessageBox()` instead of `wxLog()`.

> I guess the link errors in the CI builds are due to bakefile not having been rerun after the changes in the sample bkl file?

Yes, I did not rebake the files. The bakefile version available for download does not support GCC response files and I failed to build the latest version from the source. Its docs says I must build bakefile with the same version of MSVC the Python version it uses was built with, this may not be possible for me. I could try that again during weekend but I cannot promise a successful result.

PBfordev commented on samples/event/chessboard.h:
```diff
@@ -0,0 +1,90 @@
+// @TODO: what to write here, the info in the docs seem to be incorrect?
+typedef void (wxEvtHandler::* ChessBoardEventFunction)(ChessBoardEvent&);
+#define ChessBoardEventHandler(func) wxEVENT_HANDLER_CAST(ChessBoardEventFunction, func)
+
```
I wonder what the comment here should be. The code in docs reads

> // if you want to support old compilers you need to use some ugly macros:
> typedef void (wxEvtHandler::*MyPlotEventFunction)(MyPlotEvent&);
> #define MyPlotEventHandler(func) wxEVENT_HANDLER_CAST(MyPlotEventFunction, func)
> // if your code is only built using reasonably modern compilers, you could just
> // do this instead:
> #define MyPlotEventHandler(func) (&func)

but vadz told me that the part about old compilers is not true.

vadz replied:
Sorry, looking at this closer it seems that `wxEVENT_HANDLER_CAST` is needed, i.e. I think that if you remove it, the code won't compile even with recent gcc.

PBfordev replied:
But what should we actually write there in comments? Also, if I understand the macros correctly, these are needed only for event tables and `Connect()` and not when only` Bind()` is used?

vadz replied:
Yes, exactly. So the comment should say something like this:
```
// If the new event class is to be used with event table macros or the legacy Connect(),
// this macro must be used to cast the type of a function handling it to the type
// expected by the event table machinery, see its use in EVT_XXX below.
//
// Note that you can skip defining this macro entirely if you only plan to use Bind().
```

vadz committed:
Rebake after adding samples/event/chessboard.cpp

vadz commented:
I've rebaked the files, so this problem is taken care of.

Personally I _hate_ message boxes as the user, but to each their own, I guess.

PBfordev reviewed (commented):
Thanks for the rebake!

As for using `wxMessageBox()` with the new event, I do not insist on using it, I was just saying why I think it was the best considering the current state of the sample. I.e., `wxLogMessage()` and `wxMessageBox()` are used in the main frame, a custom log target using `wxLogTextCtrl` in the gestures frame... I can just replace `wxMessageBox()` in the new code with `wxLogMessage()`. But perhaps it would be best to unify event reporting across the whole sample. 

I am stuck at the macro description, see my comment above. I have also noticed some minor whitespace issues which I will take care of when the PR is going to be changed.

PBfordev committed:
Make some more improvements

Replace a TODO description with an actual
description of handler casting typedef and macro.

Replace wxMessageBox() with wxLogMessage().

Fix minor whitespace issues.

PBfordev committed:
Merge branch 'event-sample-add-new-event-class' of https://github.com/PBfordev/wxWidgets into event-sample-add-new-event-class

PBfordev commented:
I have added the comment for handler casting to the header, I dared to reword your proposed one somehow.

I have also replaced `wxMessageBox()` with `wxLogMessage()` and fixed the whitespace issue.

vadz commented:
Just FYI: CI failures seem to be bogus.

vadz commented on samples/event/chessboard.h:
```diff
@@ -0,0 +1,93 @@
+};
+
+
+// Declare new events,
```
I think we should be more explicit:
```suggestion
// Declare new event types,
```
IME there is a lot of confusion between event classes and types (with IDs sometimes thrown in for good measure), so being more clear shouldn't hurt.

PBfordev replied:
Done.

vadz commented on samples/event/chessboard.h:
```diff
@@ -0,0 +1,93 @@
+
+// These defines are needed only if the new event is to be used with
+// event tables, i.e., when only Bind() is going to be used you do not need them.
+#define etEVT_CHESSBOARD_CLICKED(id, func) \
```
Sorry for a maybe stupid question, but what does `et` prefix mean? Shouldn't we use `my` here, as for the classes?

But if we wanted to be parallel with the standard event types, we'd actually use `myEVT_XXX` for the event types themselves above and just `EVT_XXX` for the macros here. Not sure if it's really necessary, but it does seem a bit confusing to do exactly the converse with the prefixes than what we do in wx...

PBfordev replied:
"et" was meant as a prefix for **e**vent **t**able macros. I wanted to keep the event type "clean" and have perhaps less used legacy macro names with a prefix, failing to realize that I contradict the wxWidgets style here.

Thanks for pointing this out, I will change it: add "my" prefix to the event types and remove the prefix from the event table macros.

vadz reviewed (commented):
Thanks, this looks nice to me (much more ambitious than what I thought it would be, but I guess a realistic example is better than a completely artificial one) and I'm ready to merge this, except if you'd like to change the prefixes as discussed in the other comment. Please let me know, TIA!

PBfordev committed:
Implement suggestion from PR review

Also remove useless call to Proportion()
when adding the chessboard to the dialog.

PBfordev commented:
I have made changes requested in the review.

If this is sample is to be cleaned up[1], it should be in a separate PR.

[1] What I noticed while making changes (and mostly already commented on):

- Reporting the events to the user should be unified, there are now three ways of doing that. Perhaps having `wxLogTextCtrl `take the lower middle part of the bottom frame would be for the best. Perhaps even some message now displayed in the status bar (they are mostly cut off for me unless I resize the frame) should go the log as well.
- Statement that  "any class wishing to process wxWidgets events must use wxDECLARE_EVENT_TABLE macro" should be removed from the MyFrame declaration.
- `OnConnect()` should be removed, it does not even use `Connect()`, its code is exactly the same as in `OnBind()`.
- Comment for `wxIMPLEMENT_APP()` could be improved. I would not even talk about using a static object for `wxApp` here.

I have not checked if the event processing step numbers still match the text in the docs they refer to, hopefully they do.

vadz commented:
Thanks again for your work on this! The only small improvement to the sample functionality I'd consider would be to change the mouse cursor while dragging, as otherwise it's not really obvious that a drag is in progress. But it's not really indispensable.

Of course, if you'd like to fix any of the issues mentioned above, this would be very welcome too, samples definitely suffer from lack of attention as most of them are never changed, and stuff just keeps accreeting without ever being cleaned up.
